Sources/Logto/Core/LogtoCore.swift

```diff
@@ -54,4 +54,32 @@ enum LogtoCore {
 
         return url
     }
+
+    static func verifyAndParseSignInCallbackUri(_ callbackUri: String, redirectUri: String,
+                                                state: String) throws -> String
+    {
+        guard callbackUri.starts(with: redirectUri) else {
+            throw LogtoErrors.UriVerification.redirectUriMismatched
+        }
+
+        guard let components = URLComponents(string: callbackUri) else {
+            throw LogtoErrors.UriVerification.decodeComponentsFailed
+        }
+
+        guard components.queryItems?.contains(where: { $0.name == "state" && $0.value == state }) ?? false else {
+            throw LogtoErrors.UriVerification.stateMismatched
+        }
+
+        let errorItems = components.queryItems?.filter { ["error", "error_description"].contains($0.name) } ?? []
+
+        guard errorItems.count == 0 else {
+            throw LogtoErrors.UriVerification.errorItemFound(items: errorItems)
+        }
+
+        guard let code = components.queryItems?.first(where: { $0.name == "code" })?.value else {
```

Comment:
interesting point. What if more than one code occurred in the URL 😆 . In js, we use `query-string` to parse the query params. The return type of code was hardcoded as `string`, which in fact, if more than one code are present, will return an array.  cc: @wangsijie @IceHe should we worry about that?

Reply:
what's the case that provider will return more than one code?

Reply:
if return an array, it seems to be a really bad sign —— server has something wrong or is insecure.

A. throw exception
B. select first or last element in the array

I prefer to A

Reply:
discussed offline, keep as it is

Reply:
No need to deal with uri containing more than one `code`.

If more than one, fail by default and keep implementation simple.
We can fix it when needed someday.

Or we may need to concern all query params, which should be a string and maybe an array

Reply:
although we do not fail here - the principle is to keep it simple to when parsing query items in SDKs. whether use `.first` or recognize as a single string